gkr_iop/src/gkr/layer/zerocheck_layer.rs

```diff
@@ -41,36 +43,90 @@ pub(crate) struct RotationClaims<E: ExtensionField> {
 }
 
 pub trait ZerocheckLayer<E: ExtensionField> {
+    // static expression on circuit setup
+    fn build_static_expression(&mut self);
+
     #[allow(clippy::too_many_arguments)]
     fn prove<PB: ProverBackend<E = E>, PD: ProverDevice<PB>>(
         &self,
         num_threads: usize,
         max_num_variables: usize,
         wit: LayerWitness<PB>,
         out_points: &[Point<PB::E>],
+        pub_io_evals: &[PB::E],
         challenges: &[PB::E],
         transcript: &mut impl Transcript<PB::E>,
         num_instances: usize,
     ) -> (LayerProof<PB::E>, Point<PB::E>);
 
+    #[allow(clippy::too_many_arguments)]
     fn verify(
         &self,
         max_num_variables: usize,
         proof: LayerProof<E>,
         eval_and_dedup_points: Vec<(Vec<E>, Option<Point<E>>)>,
+        pub_io_evals: &[E],
         challenges: &[E],
         transcript: &mut impl Transcript<E>,
         num_instances: usize,
     ) -> Result<LayerClaims<E>, BackendError>;
 }
 
 impl<E: ExtensionField> ZerocheckLayer<E> for Layer<E> {
+    fn build_static_expression(&mut self) {
+        let span = entered_span!("gen_expr", profiling_4 = true);
+        // build rotation expression
+        let num_rotations = self.rotation_exprs.1.len();
+        let rotation_expr = if num_rotations > 0 {
+            let alpha_pows_expr = (2..)
+                .take(num_rotations)
+                .map(|id| Expression::Challenge(id as ChallengeId, 1, E::ONE, E::ZERO))
+                .collect_vec();
+            let rotation_expr = (0..)
+                .tuples()
+                .take(num_rotations)
+                .zip_eq(&alpha_pows_expr)
+                .map(|((rotate_wit_id, target_wit_id), alpha)| {
+                    alpha * (Expression::WitIn(rotate_wit_id) - Expression::WitIn(target_wit_id))
+                })
+                .sum::<Expression<E>>();
+            let rotation_selector_expr = Expression::<E>::WitIn((num_rotations * 2) as WitnessId);
+            Some(rotation_selector_expr * rotation_expr)
+        } else {
+            None
+        };
+
+        // build main sumcheck expression
+        let alpha_pows_expr = (2..)
+            .take(self.exprs.len() + num_rotations * ROTATION_OPENING_COUNT)
+            .map(|id| Expression::Challenge(id as ChallengeId, 1, E::ONE, E::ZERO))
+            .collect_vec();
```

Comment:
Is `Expression::Challenge(id as ChallengeId, 1, E::ONE, E::ZERO)` or `Expression::Challenge(1, id, E::ONE, E::ZERO)`?